ddl/ddl_tiflash_api.go

```diff
@@ -463,6 +511,125 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
 	return allReplicaReady, nil
 }
 
+func (d *ddl) pollTiFlashPeerInfo(ctx sessionctx.Context, pollTiFlashContext *TiFlashManagementContext) (bool, error) {
+	allReplicaReady := true
+
+	if len(pollTiFlashContext.TiFlashStores) == 0 {
+		return true, nil
+	}
+
+	// Start to process every table.
+	schema := d.GetInfoSchemaWithInterceptor(ctx)
+	if schema == nil {
+		return false, errors.New("Schema is nil")
+	}
+
+	var tableMap = make(map[int64]TiFlashReplicaStatus)
+	// Collect TiFlash Replica info, for every table.
+	for _, db := range schema.AllSchemas() {
+		tbls := schema.SchemaTables(db.Name)
+		for _, tbl := range tbls {
+			tblInfo := tbl.Meta()
+			GetTiFlashReplicaMapInfo(tblInfo, &tableMap)
+		}
+	}
+
+	tiflashStoreIDs := make([]int64, len(pollTiFlashContext.TiFlashStores))
+
+	for i, store := range pollTiFlashContext.TiFlashStores {
+		tiflashStoreIDs[i-1] = store.Store.ID
+	}
+
+	sort.Slice(tiflashStoreIDs, func(i, j int) bool { return tiflashStoreIDs[i] < tiflashStoreIDs[j] })
+
+	isTiflashPeer := func(tiflashPeerIDs []int64, peerID int64) bool {
+		for _, tiflashPeerID := range tiflashPeerIDs {
+			if tiflashPeerID == peerID {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	isTiflashAllPeerDown := func(tiflashPeerIDs []int64, tiflashDownPeerIDs []int64) bool {
+		return reflect.DeepEqual(tiflashPeerIDs, tiflashDownPeerIDs)
+	}
+
+	removeDuplicate := func(peerIDs []int64) []int64 {
+		result := make([]int64, 0, len(peerIDs))
+		temp := map[int64]struct{}{}
+		for _, item := range peerIDs {
+			if _, ok := temp[item]; !ok {
+				temp[item] = struct{}{}
+				result = append(result, item)
+			}
+		}
+		return result
+	}
+
+	var invaildRegions helper.RegionsInfo
+
+	if err := infosync.GetTiFlashPDInvalidRegionsInfo(context.Background(), &invaildRegions); err != nil {
+		return false, errors.Trace(err)
+	}
+	// There are no down-peer/pending-peer
+	if invaildRegions.Count == 0 {
+		return true, nil
+	}
+
+	var invalidTableIDs []int64
+	for _, invaildRegion := range invaildRegions.Regions {
+		var invalidPeerIDs []int64
+		for _, downPeer := range invaildRegion.DownPeers {
+			if isTiflashPeer(tiflashStoreIDs, downPeer.Peer.ID) {
+				invalidPeerIDs = append(invalidPeerIDs, downPeer.Peer.ID)
+			}
+		}
+
+		for _, pendingPeer := range invaildRegion.PendingPeers {
+			if isTiflashPeer(tiflashStoreIDs, pendingPeer.ID) {
+				invalidPeerIDs = append(invalidPeerIDs, pendingPeer.ID)
+			}
+		}
+
+		if len(invalidPeerIDs) != 0 {
+			sort.Slice(invalidPeerIDs, func(i, j int) bool { return invalidPeerIDs[i] < invalidPeerIDs[j] })
+
+			invalidPeerIDs = removeDuplicate(invalidPeerIDs)
+			if isTiflashAllPeerDown(tiflashStoreIDs, invalidPeerIDs) {
+				invaildTableID := helper.GetTiFlashTableIDFromEndKey(invaildRegion.EndKey)
+				invalidTableIDs = append(invalidTableIDs, invaildTableID)
+			}
```

Comment:
I think we can use an `invalidTableMap`. If we found any region of a table is invalid, then we can skip all regions of the table.

Reply:
I don't get it. cause we always need loop all of `invaildRegions` to compare with all of the `tableIDs`.

Reply:
for each region with invalid status
  if (the table id of the region contained by `invalidTableMap`)
    next region
  if (the region is in by a table with tiflash replica && the region's tiflash peer is all invalid)
    add the table id to `invalidTableMap`
end

Reply:
Changed this part of logic.